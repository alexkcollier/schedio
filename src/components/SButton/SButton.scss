.button {
  @include button-base;

  &--size-small {
    @include button-size-small;
  }

  &--size-regular {
    @include button-size-regular;
  }

  &--size-large {
    @include button-size-large;
  }

  &--loading {
    color: transparent !important;
    pointer-events: none;
    position: relative;
    text-shadow: none !important;
    user-select: none;

    .icon {
      filter: none;
      stroke: transparent !important;
    }

    .button__spinner {
      left: 50%;
      padding-top: spacing('half');
      position: absolute;
      top: 0;
      transform: translateX(-50%);
    }
  }

  &--icon-only {
    padding: spacing('half');

    .icon {
      margin: 0;

      & + .icon {
        margin-left: spacing('half');
      }
    }
  }

  &__icon {
    &--left {
      margin-left: -1 * spacing('quarter');
      margin-right: spacing('tight');
    }

    &--right {
      margin-left: spacing('tight');
      margin-right: -1 * spacing('quarter');
    }

    .button:disabled & {
      filter: none;
      stroke: color('night', 'lightest');
    }
  }

  @each $color-name in $button-colors {
    &--color-#{$color-name} {
      @include button-color($color-name);

      &-outlined {
        @include button-outline($color-name);
      }

      &-text {
        @include button-text($color-name);
      }
    }
  }
}
