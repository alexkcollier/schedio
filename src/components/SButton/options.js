export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options'

export const sizes = [
  'small',
  'regular',
  'large',
]

export const types = [
  'outlined',
  'text',
]
