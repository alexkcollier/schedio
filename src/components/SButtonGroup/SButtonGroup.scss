.button-group {
  .button {
    margin-right: spacing();
    margin-bottom: spacing();
  }

  @each $color-name in $button-colors {
    &--color-#{$color-name} {
      .button {
        @include button-color($color-name);
      }
    }

    &--color-#{$color-name}-outlined {
      .button {
        @include button-outline($color-name);
      }
    }
  }

  &--grouped {
    border-radius: border-radius('small');
    box-shadow: shadow();
    display: inline-block;

    .button {
      border-radius: 0;
      box-shadow: none;
      margin: 0;
      position: relative;

      &:first-child {
        border-top-left-radius: border-radius('small');
        border-bottom-left-radius: border-radius('small');
      }

      &:last-child {
        border-top-right-radius: border-radius('small');
        border-bottom-right-radius: border-radius('small');
      }

      &:not(:last-child) {
        border-right: 0;
      }

      &:focus {
        // Ensures outline fully visible when focused
        z-index: 10;
      }
    }

    @each $color-name in $button-colors {
      &.button-group--color-#{$color-name} {
        .button {
          &:not(:first-child) {
            // Ensures borders between buttons in grouped button groups are darkened
            // relative to their background
            border-left-color: darken(color($color-name), 20%);
          }
        }
      }
    }
  }

  &--disabled {
    box-shadow: none;
  }
}
