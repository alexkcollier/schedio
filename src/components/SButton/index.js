import SButton from './SButton.vue';

export { SButton };

export default { install: Vue => Vue.component(SButton.name, SButton) };
