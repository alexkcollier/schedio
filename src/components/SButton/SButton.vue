<template>
  <button
    v-bind="$attrs"
    :class="classList"
    class="button"
    v-on="$listeners"
  >
    <SIcon
      v-if="iconLeft"
      :icon="iconLeft"
      class="button__icon button__icon--left"
    />

    <slot v-if="!iconOnly" />

    <SIcon
      v-if="iconRight"
      :icon="iconRight"
      class="button__icon button__icon--right"
    />
  </button>
</template>

<script>
import { colors, sizes } from './options.js'

export default {
  name: 'SButton',

  props: {
    buttonColor: {
      type: String,
      default: '',
      validator: value => {
        if (!value || colors.includes(value)) return true

        return console.error(`\`buttonColor\` ${value} not found. Allowed colors: ${[...colors]}`)
      }
    },

    buttonSize: {
      type: String,
      default: '',
      validator: value => {
        if (!value || sizes.includes(value)) return true

        return console.error(`\`buttonSizes\` ${value} not found. Allowed sizes: ${sizes}`)
      }
    },

    iconOnly: {
      type: Boolean,
      default: false
    },

    iconLeft: {
      type: String,
      default: ''
    },

    iconRight: {
      type: String,
      default: ''
    }
  },

  computed: {
    classList() {
      return [
        // Handles storybook default
        this.buttonColor ? `button--color-${this.buttonColor}` : 'button--color-green',
        this.buttonSize && `button--size-${this.buttonSize}`,
        this.iconOnly && `button--icon-only`
      ]
    }
  }
}
</script>
