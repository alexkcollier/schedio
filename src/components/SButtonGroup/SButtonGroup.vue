<template>
  <div
    :class="classList"
    class="button-group"
  >
    <slot />
  </div>
</template>

<script>
import { colorNames } from '@/components/SButton/options.js'

export default {
  name: 'SButtonGroup',

  props: {
    color: {
      type: String,
      default: '',
      validator: (value) => {
        if (!value || colorNames.includes(value)) return true
        return console.error(`\`color\` ${value} not found. Allowed colors: ${colorNames}`)
      },
    },

    isOutlined: {
      type: Boolean,
      default: false,
    },

    isText: {
      type: Boolean,
      default: false,
    },

    isGrouped: {
      type: Boolean,
      default: false,
    },
  },

  data () {
    return {
      isDisabled: false,
    }
  },

  computed: {
    classList () {
      return [
        this.groupStyle,
        {
          'button-group--grouped': this.isGrouped,
          'button-group--disabled': this.isDisabled,
        },
      ]
    },
    groupStyle () {
      let style = ''

      if (this.color) style += `button-group--color-${this.color}`

      if (style && this.isOutlined) style += '-outlined'

      if (style && this.isText) style += '-text'

      return style
    },
  },

  updated () {
    const buttons = this.$children.filter(child => child.$options.name === 'SButton')
    this.isDisabled = buttons.every(button => !!button.$attrs.disabled)
  },
}
</script>

<style lang="scss">
.button-group {
  margin-bottom: -1 * spacing();
  margin-right: -1 * spacing();

  .button {
    margin-bottom: spacing();
    margin-right: spacing();
  }

  @each $color-name in $button-colors {
    &--color-#{$color-name} {
      .button {
        @include button-color($color-name);
      }
    }

    &--color-#{$color-name}-text {
      box-shadow: none !important;

      .button {
        @include button-text($color-name);
      }
    }

    &--color-#{$color-name}-outlined {
      .button {
        @include button-outline($color-name);
      }
    }
  }

  &--grouped {
    border-radius: border-radius('small');
    box-shadow: shadow();
    display: inline-block;
    margin: 0;

    .button {
      border-radius: 0;
      box-shadow: none;
      margin: 0;
      position: relative;

      &:first-child {
        border-bottom-left-radius: border-radius('small');
        border-top-left-radius: border-radius('small');
      }

      &:last-child {
        border-bottom-right-radius: border-radius('small');
        border-top-right-radius: border-radius('small');
      }

      &:not(:last-child) {
        border-right: 0;
      }

      &:focus {
        // Ensures outline fully visible when focused
        z-index: 10;
      }
    }

    @each $color-name in $button-colors {
      &.button-group--color-#{$color-name} {
        .button {
          &:not(:first-child) {
            // Ensures borders between buttons in grouped button groups are darkened
            // relative to their background
            border-left-color: darken(color($color-name), 20%);
          }
        }
      }
    }
  }

  &--disabled {
    box-shadow: none;
  }
}
</style>
