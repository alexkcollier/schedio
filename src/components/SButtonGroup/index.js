import SButtonGroup from './SButtonGroup.vue'

export { SButtonGroup }

export default { install: Vue => Vue.component(SButtonGroup.name, SButtonGroup) }
