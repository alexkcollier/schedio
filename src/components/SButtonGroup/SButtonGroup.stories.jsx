import { SButton } from '@/components/SButton'
import { colors } from '@/components/SButton/options'
import { SHeading } from '@/components/SHeading'
import { SButtonGroup } from '@/components/SButtonGroup'
import PropList from '@/docs/PropList'
import { boolean, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

storiesOf('Components/Buttons.SButtonGroup', module)
  .addParameters({ jest: 'SButtonGroup' })
  .add('Button Group', () => {
    return {
      props: {
        groupColor: { default: select('group-color', ['', ...colors], '', 'Optional Props') },
        groupOutlineColor: {
          default: select('group-outline-color', ['', ...colors], '', 'Optional Props')
        },
        isGrouped: { default: boolean('is-grouped', false, 'Optional Props') },
        button1Text: { default: text('Button 1 text', 'Button 1', 'Slots') },
        button2Text: { default: text('Button 2 text', 'Button 2', 'Slots') },
        button3Text: { default: text('Button 3 text', 'Button 3', 'Slots') },
        button1Disabled: { default: boolean('Button 1 disabled', false, '$attrs') },
        button2Disabled: { default: boolean('Button 2 disabled', false, '$attrs') },
        button3Disabled: { default: boolean('Button 3 disabled', false, '$attrs') }
      },
      render(h) {
        return (
          <div>
            <SHeading level="1">Button Group</SHeading>
            <p>Button groups take multiple buttons and display them nicely.</p>
            <p>
              Buttons in a group can be merged with the <code>isGrouped</code> prop.
            </p>
            <p>Note that the button group's color will override an individual button's color.</p>

            <SHeading level="2">Example</SHeading>
            <SButtonGroup
              groupColor={this.groupColor}
              groupOutlineColor={this.groupOutlineColor}
              isGrouped={this.isGrouped}
            >
              <SButton disabled={this.button1Disabled}>{this.button1Text}</SButton>
              <SButton disabled={this.button2Disabled}>{this.button2Text}</SButton>
              <SButton disabled={this.button3Disabled}>{this.button3Text}</SButton>
            </SButtonGroup>

            <PropList component={SButtonGroup} />
          </div>
        )
      }
    }
  })
