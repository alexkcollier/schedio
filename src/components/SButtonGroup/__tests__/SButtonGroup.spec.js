import { mount } from '@vue/test-utils'
import SButton from '@/components/SButton/SButton.vue'
import SButtonGroup from '@/components/SButtonGroup/SButtonGroup.vue'

describe('SButtonGroup.vue', () => {
  let wrapper
  let buttonGroup

  beforeEach(() => {
    wrapper = mount(
      {
        components: { SButton, SButtonGroup },
        template: `
          <SButtonGroup>
            <SButton v-for="i in 3" :key="i">
              Button {{ i }}
            </SButton>
          </SButtonGroup>
        `
      },
      { stubs: { SButton: true } }
    )

    buttonGroup = wrapper.find(SButtonGroup)
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders child buttons', () => {
    expect(buttonGroup.is(SButtonGroup)).toBe(true)
    expect(wrapper.contains(SButton)).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('changes child button colors', () => {
    buttonGroup.setProps({ groupColor: 'red' })
    expect(buttonGroup.contains('.button-group--color-red')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('groups child buttons', () => {
    buttonGroup.setProps({ areGrouped: true })
    expect(buttonGroup.contains('.button-group--grouped')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
