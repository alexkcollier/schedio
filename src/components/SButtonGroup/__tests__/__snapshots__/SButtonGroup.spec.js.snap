// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SButtonGroup.vue disables shadow based on children disabled status 1`] = `
<div class="button-group button-group--disabled"><button class="button" disabled="disabled">
    <!---->
    <!---->
    Button 1
    <!----></button><button class="button" disabled="disabled">
    <!---->
    <!---->
    Button 2
    <!----></button><button class="button" disabled="disabled">
    <!---->
    <!---->
    Button 3
    <!----></button></div>
`;

exports[`SButtonGroup.vue groups child buttons 1`] = `
<div class="button-group button-group--grouped">
  <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="">
    Button 1
  </sbutton-stub>
  <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="">
    Button 2
  </sbutton-stub>
  <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="">
    Button 3
  </sbutton-stub>
</div>
`;

exports[`SButtonGroup.vue renders child buttons 1`] = `
<div class="button-group">
  <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="">
    Button 1
  </sbutton-stub>
  <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="">
    Button 2
  </sbutton-stub>
  <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="">
    Button 3
  </sbutton-stub>
</div>
`;

exports[`SButtonGroup.vue renders correctly 1`] = `
<div class="button-group">
  <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="">
    Button 1
  </sbutton-stub>
  <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="">
    Button 2
  </sbutton-stub>
  <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="">
    Button 3
  </sbutton-stub>
</div>
`;
